File: monitor/src/main/java/com/xter/monitor/L.java
package com.xter.monitor;

import android.text.TextUtils;
import android.util.Log;

public class L {

	public static boolean DEBUG = true;

	public static void v(String msg) {
		if (!TextUtils.isEmpty(msg)) {
			String tag = getMethodPath(4, 4);
			if (DEBUG) {
				Log.v(tag, msg);
			} else {
//				writeToFile(tag + ":" + msg);
			}
		}
	}

	public static void d(String msg) {
		if (!TextUtils.isEmpty(msg)) {
			String tag = getMethodPath(4, 4);
			if (DEBUG) {
				Log.d(tag, msg);
			} else {
				writeToFile(tag + ":" + msg);
			}
		}
	}

	public static void i(String msg) {
		if (!TextUtils.isEmpty(msg)) {
			String tag = getMethodPath(4, 4);
			if (DEBUG) {
				Log.i(tag, msg);
			} else {
				writeToFile(tag + ":" + msg);
			}
		}
	}

	public static void w(String msg) {
		if (!TextUtils.isEmpty(msg)) {
			String tag = getMethodPath(4, 4);
			if (DEBUG) {
				Log.w(tag, msg);
			} else {
				writeToFile(tag + ":" + msg);
			}
		}
	}

	public static void e(String msg) {
		if (!TextUtils.isEmpty(msg)) {
			String tag = getMethodPath(4, 4);
			if (DEBUG) {
				Log.e(tag, msg);
			} else {
				writeToFile(tag + ":" + msg);
			}
		}
	}

	public static void exception(int prior, Throwable ex) {
		L.w(prior, ex.toString());
	}

	private static void writeToFile(String content) {
	}

	public static void v(String tag, String msg) {
		if (DEBUG && !TextUtils.isEmpty(msg))
			Log.v(tag, msg);
	}

	public static void d(String tag, String msg) {
		if (DEBUG && !TextUtils.isEmpty(msg))
			Log.d(tag, msg);
	}

	public static void i(String tag, String msg) {
		if (DEBUG && !TextUtils.isEmpty(msg))
			Log.i(tag, msg);
	}

	public static void w(String tag, String msg) {
		if (DEBUG && !TextUtils.isEmpty(msg))
			Log.w(tag, msg);
	}

	public static void e(String tag, String msg) {
		if (DEBUG && !TextUtils.isEmpty(msg))
			Log.e(tag, msg);
	}

	public static void v(int classPrior, int methodPriorint, String msg) {
		if (DEBUG && !TextUtils.isEmpty(msg))
			Log.v(getMethodPath(classPrior, methodPriorint), msg);
	}

	public static void d(int classPrior, int methodPrior, String msg) {
		if (DEBUG && !TextUtils.isEmpty(msg))
			Log.d(getMethodPath(classPrior, methodPrior), msg);
	}

	public static void i(int classPrior, int methodPrior, String msg) {
		if (DEBUG && !TextUtils.isEmpty(msg))
			Log.i(getMethodPath(classPrior, methodPrior), msg);
	}

	public static void w(int classPrior, int methodPrior, String msg) {
		if (DEBUG && !TextUtils.isEmpty(msg))
			Log.w(getMethodPath(classPrior, methodPrior), msg);
	}

	public static void e(int classPrior, int methodPrior, String msg) {
		if (DEBUG && !TextUtils.isEmpty(msg))
			Log.e(getMethodPath(classPrior, methodPrior), msg);
	}

	public static void v(int prior, String msg) {
		if (DEBUG && !TextUtils.isEmpty(msg))
			Log.v(getMethodPath(prior, prior), msg);
	}

	public static void d(int prior, String msg) {
		if (DEBUG && !TextUtils.isEmpty(msg))
			Log.d(getMethodPath(prior, prior), msg);
	}

	public static void i(int prior, String msg) {
		if (DEBUG && !TextUtils.isEmpty(msg))
			Log.i(getMethodPath(prior, prior), msg);
	}

	public static void w(int prior, String msg) {
		if (DEBUG && !TextUtils.isEmpty(msg))
			Log.w(getMethodPath(prior, prior), msg);
	}

	public static void e(int prior, String msg) {
		if (DEBUG && !TextUtils.isEmpty(msg))
			Log.e(getMethodPath(prior, prior), msg);
	}


	/**
	 * 得到调用此方法的类名与方法名
	 *
	 * @param classPrior  类级
	 * @param methodPrior 方法级
	 * @return string
	 */
	public static String getMethodPath(int classPrior, int methodPrior) {
		StackTraceElement[] stackTrace = Thread.currentThread().getStackTrace();

		StackTraceElement targetElement = stackTrace[classPrior];
		String fileName = targetElement.getFileName();
//		String className = targetElement.getClassName();
//		String[] classNameInfo = className.split("\\.");
//		if (classNameInfo.length > 0) {
//			className = classNameInfo[classNameInfo.length - 1] + SUFFIX;
//		}
//
//		if (className.contains("$")) {
//			className = className.split("\\$")[0] + SUFFIX;
//		}

		String methodName = targetElement.getMethodName();
		int lineNumber = targetElement.getLineNumber();

		if (lineNumber < 0) {
			lineNumber = 0;
		}
		int length = Thread.currentThread().getStackTrace().length;
		if (classPrior > length || methodPrior > length) {
			return "";
		} else
			return "(" + fileName + ":" + lineNumber + ")#" + methodName + "-->";
	}

	/**
	 * 得到调用此方法的类名与方法名（带包名）
	 *
	 * @param classPrior  类级
	 * @param methodPrior 方法级
	 * @return string
	 */
	public static String getPackageMethodPath(int classPrior, int methodPrior) {
		StackTraceElement[] stackTrace = Thread.currentThread().getStackTrace();

		StackTraceElement targetElement = stackTrace[classPrior];
		String className = targetElement.getClassName();
		String methodName = targetElement.getMethodName();
		int lineNumber = targetElement.getLineNumber();

		if (lineNumber < 0) {
			lineNumber = 0;
		}
		int length = Thread.currentThread().getStackTrace().length;
		if (classPrior > length || methodPrior > length) {
			return "";
		} else
			return "(" + className + ":" + lineNumber + ")#" + methodName + "-->";
	}

	/**
	 * 测试方法，将线程中的序列全部输出
	 */
	public static void logThreadSequence() {
		int length = Thread.currentThread().getStackTrace().length;
		for (int i = 0; i < length; i++) {
			Log.i(Thread.currentThread().getStackTrace()[i].getClassName(),
					Thread.currentThread().getStackTrace()[i].getMethodName());
		}
	}

}

File: app/src/main/java/com/xter/slimnotek/util/L.kt
package com.xter.slimnotek.util

import android.text.TextUtils
import android.util.Log
import com.jinxin.gateway3.common.util.LogFileWriter

/**
 * 日志工具类
 */
object L {

    /**
     * 调试标志，为true则显示实时日志，为false则写入日志文件
     */
    var DEBUG: Boolean = false

    fun v(msg: String) {
        if (!TextUtils.isEmpty(msg)) {
            val tag = getMethodPath(4, 4)
            if (DEBUG) {
                Log.v(tag, msg)
            } else {
                //				writeToFile(tag + ":" + msg);
            }
        }
    }

    fun d(msg: String) {
        if (!TextUtils.isEmpty(msg)) {
            val tag = getMethodPath(4, 4)
            if (DEBUG) {
                Log.d(tag, msg)
            } else {
                writeToFile("$tag:$msg")
            }
        }
    }

    fun i(msg: String) {
        if (!TextUtils.isEmpty(msg)) {
            val tag = getMethodPath(4, 4)
            if (DEBUG) {
                Log.i(tag, msg)
            } else {
                writeToFile("$tag:$msg")
            }
        }
    }

    fun w(msg: String) {
        if (!TextUtils.isEmpty(msg)) {
            val tag = getMethodPath(4, 4)
            if (DEBUG) {
                Log.w(tag, msg)
            } else {
                writeToFile("$tag:$msg")
            }
        }
    }

    fun e(msg: String) {
        if (!TextUtils.isEmpty(msg)) {
            val tag = getMethodPath(4, 4)
            if (DEBUG) {
                Log.e(tag, msg)
            } else {
                writeToFile("$tag:$msg")
            }
        }
    }

    private fun writeToFile(content: String) {
        LogFileWriter.instance.write(content + "\r\n")
    }

    fun v(tag: String, msg: String) {
        if (DEBUG && !TextUtils.isEmpty(msg))
            Log.v(tag, msg)
    }

    fun d(tag: String, msg: String) {
        if (DEBUG && !TextUtils.isEmpty(msg))
            Log.d(tag, msg)
    }

    fun i(tag: String, msg: String) {
        if (DEBUG && !TextUtils.isEmpty(msg))
            Log.i(tag, msg)
    }

    fun w(tag: String, msg: String) {
        if (DEBUG && !TextUtils.isEmpty(msg))
            Log.w(tag, msg)
    }

    fun e(tag: String, msg: String) {
        if (DEBUG && !TextUtils.isEmpty(msg))
            Log.e(tag, msg)
    }

    fun v(classPrior: Int, methodPriorint: Int, msg: String) {
        if (DEBUG && !TextUtils.isEmpty(msg))
            Log.v(getMethodPath(classPrior, methodPriorint), msg)
    }

    fun d(classPrior: Int, methodPrior: Int, msg: String) {
        if (DEBUG && !TextUtils.isEmpty(msg))
            Log.d(getMethodPath(classPrior, methodPrior), msg)
    }

    fun i(classPrior: Int, methodPrior: Int, msg: String) {
        if (DEBUG && !TextUtils.isEmpty(msg))
            Log.i(getMethodPath(classPrior, methodPrior), msg)
    }

    fun w(classPrior: Int, methodPrior: Int, msg: String) {
        if (DEBUG && !TextUtils.isEmpty(msg))
            Log.w(getMethodPath(classPrior, methodPrior), msg)
    }

    fun e(classPrior: Int, methodPrior: Int, msg: String) {
        if (DEBUG && !TextUtils.isEmpty(msg))
            Log.e(getMethodPath(classPrior, methodPrior), msg)
    }

    fun v(prior: Int, msg: String) {
        if (DEBUG && !TextUtils.isEmpty(msg))
            Log.v(getMethodPath(prior, prior), msg)
    }

    fun d(prior: Int, msg: String) {
        if (!TextUtils.isEmpty(msg)) {
            val tag = getMethodPath(prior, prior)
            if (DEBUG) {
                Log.d(getMethodPath(prior, prior), msg)
            } else {
                writeToFile("$tag:$msg")
            }
        }
    }

    fun i(prior: Int, msg: String) {
        if (!TextUtils.isEmpty(msg)) {
            val tag = getMethodPath(prior, prior)
            if (DEBUG) {
                Log.i(getMethodPath(prior, prior), msg)
            } else {
                writeToFile("$tag:$msg")
            }
        }
    }

    fun w(prior: Int, msg: String) {
        if (!TextUtils.isEmpty(msg)) {
            val tag = getMethodPath(prior, prior)
            if (DEBUG) {
                Log.w(getMethodPath(prior, prior), msg)
            } else {
                writeToFile("$tag:$msg")
            }
        }
    }

    fun e(prior: Int, msg: String) {
        if (!TextUtils.isEmpty(msg)) {
            val tag = getMethodPath(prior, prior)
            if (DEBUG) {
                Log.e(getMethodPath(prior, prior), msg)
            } else {
                writeToFile("$tag:$msg")
            }
        }
    }

    fun exception(prior: Int, ex: Throwable) {
        w(prior, ex.toString())
        //		L.w(prior, track(ex));
    }

    private fun track(e: Throwable?): String? {
        val sb = StringBuffer()
        if (e != null) {
            for (element in e.stackTrace) {
                sb.append(element).append("\r\n\t")
            }
            if (sb.length > 1) {
                sb.deleteCharAt(sb.length - 1)
            }
        }
        return if (sb.length == 0) null else sb.toString()
    }

    /**
     * 得到调用此方法的类名与方法名
     *
     * @param classPrior  类级
     * @param methodPrior 方法级
     * @return string
     */
    fun getMethodPath(classPrior: Int, methodPrior: Int): String {
        val stackTrace = Thread.currentThread().stackTrace

        val targetElement = stackTrace[classPrior]
        val fileName = targetElement.fileName

        val methodName = targetElement.methodName
        var lineNumber = targetElement.lineNumber

        if (lineNumber < 0) {
            lineNumber = 0
        }
        val length = Thread.currentThread().stackTrace.size
        return if (classPrior > length || methodPrior > length) {
            ""
        } else
            "($fileName:$lineNumber)#$methodName-->"
    }

    /**
     * 得到调用此方法的类名与方法名（带包名）
     *
     * @param classPrior  类级
     * @param methodPrior 方法级
     * @return string
     */
    fun getPackageMethodPath(classPrior: Int, methodPrior: Int): String {
        val stackTrace = Thread.currentThread().stackTrace

        val targetElement = stackTrace[classPrior]
        val className = targetElement.className
        val methodName = targetElement.methodName
        var lineNumber = targetElement.lineNumber

        if (lineNumber < 0) {
            lineNumber = 0
        }
        val length = Thread.currentThread().stackTrace.size
        return if (classPrior > length || methodPrior > length) {
            ""
        } else
            "($className:$lineNumber)#$methodName-->"
    }

    /**
     * 测试方法，将线程中的序列全部输出
     */
    fun logThreadSequence() {
        val length = Thread.currentThread().stackTrace.size
        for (i in 0 until length) {
            Log.i(
                Thread.currentThread().stackTrace[i].className,
                Thread.currentThread().stackTrace[i].methodName
            )
        }
    }

}
